src/util.js

```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
+      return encodeCache.cache(val, () => getCustomRouterDetails(val))
+    } else if (val._isVue) {
+      return encodeCache.cache(val, () => getCustomInstanceDetails(val))
+    } else if (typeof val.render === 'function') {
+      return encodeCache.cache(val, () => getCustomComponentDefinitionDetails(val))
+    }
   } else if (Number.isNaN(val)) {
     return NAN
-  } else if (val instanceof RegExp) {
-    // special handling of native type
-    return `[native RegExp ${val.toString()}]`
-  } else if (val instanceof Date) {
-    return `[native Date ${val.toString()}]`
-  } else if (val && val._isVue) {
-    return getCustomInstanceDetails(val)
+  }
+  return sanitize(val)
+}
+
+export function getCustomMapDetails (val) {
+  const list = []
+  val.forEach(
+    (value, key) => list.push({
+      key,
+      value
+    })
+  )
+  return {
+    _custom: {
+      type: 'map',
+      display: 'Map',
+      value: list,
+      readOnly: true,
+      fields: {
+        abstract: true
+      }
+    }
+  }
+}
+
+export function reviveMap (val) {
+  const result = new Map()
+  const list = val._custom.value
+  for (let i = 0; i < list.length; i++) {
+    const { key, value } = list[i]
+    result.set(key, reviver(null, value))
+  }
+  return result
+}
+
+export function getCustomSetDetails (val) {
+  const list = Array.from(val)
+  return {
+    _custom: {
+      type: 'set',
+      display: `Set[${list.length}]`,
+      value: list,
+      readOnly: true
+    }
+  }
+}
+
+export function reviveSet (val) {
+  const result = new Set()
+  const list = val._custom.value
+  for (let i = 0; i < list.length; i++) {
+    const value = list[i]
+    result.add(reviver(null, value))
+  }
+  return result
+}
+
+// Use a custom basename functions instead of the shimed version
+// because it doesn't work on Windows
+function basename (filename, ext) {
+  return path.basename(
+    filename.replace(/^[a-zA-Z]:/, '').replace(/\\/g, '/'),
+    ext
+  )
+}
+
+export function getComponentName (options) {
+  const name = options.name || options._componentTag
+  if (name) {
+    return name
+  }
+  const file = options.__file // injected by vue-loader
+  if (file) {
+    return classify(basename(file, '.vue'))
+  }
+}
+
+export function getCustomComponentDefinitionDetails (def) {
+  let display = getComponentName(def)
+  if (display) {
+    if (def.name && def.__file) {
+      display += ` <span>(${def.__file})</span>`
+    }
   } else {
-    return sanitize(val)
+    display = '<i>Unknown Component</i>'
+  }
+  return {
+    _custom: {
+      type: 'component-definition',
+      display,
+      tooltip: 'Component definition',
+      ...def.__file ? {
+        file: def.__file
+      } : {}
+    }
+  }
+}
+
+export function getCustomFunctionDetails (func) {
+  const args = func.toString().match(/\(.*\)/)[0]
+  return {
+    _custom: {
+      type: 'function',
+      display: `<span>${func.name}</span>${args}`
```

Comment:
I think adding the word _function_ somewhere could make this more explicit. What about something like _ƒ o(a, b, c)_, similar to the console?